Add LED API and driver support

Mani-Sadhasivam opened the pull request:
This PR adds the following changes:

1. LED driver API support.
2. Support for relevant LED driver LP3943.
3. Sample application for the corresponding driver.
4. Reduce bus frequency and enable I2C_3 for 96Boards Neonkey

mbolivar commented on include/led.h:
```diff
@@ -4,20 +4,109 @@
+/**
+ * @brief Update an LED device frequency
+ *
+ * This routine updates the LED device frequency
```
This also could use some improvements. Is this frequency in Hertz, for example, or some other unit? What, in general, do you mean by the "frequency" of the LED device? (What is the periodic signal whose frequency is being set by this API?)

Why is it a plain signed int, instead of an unsigned type? I've only ever seen negative "frequencies" in the frequency domain representations of signals, but I don't think that's what you're referring to here.

Mani-Sadhasivam replied:
Agree. It should be unsigned and the units is cycles per second. Will make it more clear.

mbolivar commented on include/led.h:
```diff
@@ -4,20 +4,109 @@
+ * This routine updates the LED device frequency
+ *
+ * @param dev LED device
+ * @param id ID of the LED frequency register
```
This is also making assumptions that there is a single register which stores a frequency, which is also not likely to be valid on all possible drivers. Can you make this API general enough to avoid this assumption?

Mani-Sadhasivam replied:
Sure. Will make this as a conditional struct property

mbolivar commented on include/led.h:
```diff
@@ -4,20 +4,109 @@
+/**
+ * @brief Update an LED device channel
+ *
+ * This routine updates the LED device channel
```
I don't know what a "channel" means in this context. In the LED strip APIs, this is defined as either an individual RGB color channel in an addressable RGB LED (as in the led_strip_update_rgb() documentation), or otherwise one of the individually addressable LEDs that appears linearly in the strip (as in led_strip_update_channels()).

Do you mean the same thing here? If so, can you add some documentation saying so? Also, if that's what you mean, did you leave out RGB support for some reason? I can imagine people wanting to set RGB values explicitly for an individual LED (think of the LED HAL APIs on an Android phone that lets apps set a specific color in given conditions, so you can make emails set the LED red, IRC messages set it blue, etc.), so RGB support would be good to have as well.

Either way, can you try to define more clearly what is meant here? Otherwise, driver writers will get confused.

Mani-Sadhasivam replied:
Channel refers to individual LEDs here, will add more documentation. Also, I was thinking about adding RGB support later but anyhow it doesn't harm to do it now :)

mbolivar commented on include/led.h:
```diff
@@ -4,20 +4,109 @@
+ *
+ * @param dev LED device
+ * @param channel Channel number of the device
+ * @param mode Channel mode to set
```
What does a "channel mode" mean to a general purpose LED driver?

Please try to make the top-level API abstract enough to support a variety of LED devices, not just one piece of hardware.

Mani-Sadhasivam replied:
Ack.

mbolivar reviewed (changes requested):
Thanks for this. I like the idea of having a general-purpose API for LEDs, which includes LED strips. I have some questions about the structure of this API, though. (I've only looked at the patch adding the core API, not the individual drivers.)

Mani-Sadhasivam commented on include/led.h:
```diff
@@ -4,20 +4,109 @@
- *
- * This library abstracts the chipset drivers for individually
- * addressable strips of LEDs.
+ * @brief Public API for controlling LEDs
```
Makes sense! Will do it in next iteration

Mani-Sadhasivam commented on include/led.h:
```diff
@@ -4,20 +4,109 @@
+ * This routine updates the LED device duty cycle
+ *
+ * @param dev LED device
+ * @param id ID of the LED duty cycle register
```
Ok. How about having it as a conditional struct property? Like how we have `CONFIG_LED_STRIP_RGB_SCRATCH`

Mani-Sadhasivam commented on include/led.h:
```diff
@@ -4,20 +4,109 @@
+ *
+ * @param dev LED device
+ * @param id ID of the LED duty cycle register
+ * @param duty Duty cycle value
```
Will make this more clear by having `led` struct with maximum and minimum acceptable values.

Nope. Will change the type unsigned

MaureenHelm reviewed (changes requested):
This looks a lot like the pwm API. Do we really need a new API?

Mani-Sadhasivam commented:
@MaureenHelm Yeah. PWM and LED subsystems are closely tied together but differ in their purpose. Initially I thought about adding this LED driver (LP3943) under PWM subsystem but it seemed very ambiguous. Also, some SoC's have inbuilt PWM and LED subsystems which will make it more tough.

But the initial revision I have submitted didn't justify it well. Working on a hardware agnostic API which will look good.

mbolivar commented:
@Mani-Sadhasivam you might want to take a look at the lights HAL API for Android for some inspiration:

https://android.googlesource.com/platform/hardware/libhardware/+/master/include/hardware/lights.h#176

It supports a single API function called set_light(), which takes a struct light_state_t that abstracts a lot of the hardware-specific details.

I'm not saying to copy it or that it's perfect. I'm just trying to give an example of a general purpose LED API. Linux has one too, of course.

Here are some basic use cases:

- simple on/off
- brightness setting
- color setting (note the "do your best here" comment in the Android HAL documentation which describes ways to deal with a request for a color setting which can't be matched exactly, and maybe think about ways an LED driver core can provide helper functions to make "do your best" easier for driver writers)
- blinking
  - you might be able to do this in hardware if the backing LED driver uses PWM
  - you might need to schedule work to bit-bang a pin in cases where the LED is just a GPIO
  - you might be dealing with a light that's controlled via I2C like https://www.sparkfun.com/products/8579
  - "one shot" blinking is also sometimes useful (e.g. scheduling a single blink now or at some time in the future, instead of blinking the same way forever)

You might want to start with the simplest use cases first and then add additional ones. That will leave the API room to grow over time.

iha commented:
@Mani-Sadhasivam 

Would be interesting a pointer to set specific configurations of a SOC. On ESP32, for example, is possible to set a pwm fade on hardware.

Example from Kernel Linux: 
```
struct pwm_device * pwm
    PWM device

int pwm_set_chip_data(struct pwm_device * pwm, void * data)

    set private chip data for a PWM
```
https://www.kernel.org/doc/html/v4.13/driver-api/miscellaneous.html#pulse-width-modulation-pwm

Mani-Sadhasivam committed:
drivers: led: Add public API for LED drivers

Add support for LED APIs for controlling the LED devices. This
API can be used by the LED devices present on the chip and connected
externally via buses like I2C, SPI etc...

Following APIs are currently supported:

1. led_blink
2. led_set_brightness
3. led_on
4. led_off

Driver support using these APIs will be added in subsequent patches.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
drivers: led: Add LED driver support for TI LP3943

Add support for Texas Instruments LP3943 - 16 channel LED driver.
Supported blinking period: 0 to 1600ms
Supported brightness value: 0 to 100%

This driver supports the following APIs:

1. led_blink
2. led_set_brightness
3. led_on
4. led_off

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
samples: drivers: Add sample application for LP3943

Add sample application for TI LP3943 LED driver. This application
displays a continuous pattern that turns on 16 LEDs at 1s one by
one until it reaches the end and turns off LEDs in reverse order.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam committed:
boards: arm: 96b_neonkey: Fix I2C_3 for LP3943 LED driver

Following changes are necessary to use LP3943 LED driver present
in 96Boards Neonkey board

1. Enable I2C_3 interface
2. Reduce the speed of I2C_3 bus on 96Boards Neonkey. Due to some
design quirks, on board LP3943 LED driver connected to I2C_3
operates only at Standard speed of 100KHz.

Signed-off-by: Manivannan Sadhasivam <manivannan.sadhasivam@linaro.org>

Mani-Sadhasivam commented:
@mbolivar @MaureenHelm 

Pushed the new set of APIs. As per your valuable inputs, I made this API to be hardware agnostic. Current support adds following APIs:

1. led_blink
2. led_oneshot_blink
3. led_set_brightness
4. led_on
5. led_off

The relevant LED driver (LP3943) incorporates most of the APIs except `led_oneshot_blink` which I think
can be added in future.

@iha Good idea! I need to do more R&D on this before jumping in. Perhaps can be added in future too.

mbolivar commented on include/led.h:
```diff
@@ -0,0 +1,169 @@
+}
+
+/**
+ * @brief Oneshot blink of an LED
```
If this API is unimplemented, let's just leave it out for now. It can be added to the core API when there is a user.

Mani-Sadhasivam replied:
Sure, will leave it for now!

mbolivar commented on include/led.h:
```diff
@@ -0,0 +1,169 @@
+ */
+
+struct led_data {
+	/* Minimum acceptable time period (in ms) supported by the device */
```
Thanks for adding units, but I'm not sure what this means. Minimum acceptable time period for what? Please see more detailed question about duty cycle and period elsewhere.

Mani-Sadhasivam replied:
I think we can use `delay` here than using `period`. This will make the LED API to look distinct from the PWM API. Please see more below.

mbolivar commented on include/led.h:
```diff
@@ -0,0 +1,169 @@
+	u32_t min_period;
+	/* Maximum acceptable time period (in ms) supported by the device */
+	u32_t max_period;
+	/* Minimum acceptable brightness (in percent) supported by the device */
```
If this is a percentage, why is anything other than 0 to 100 chosen? The device should be able to support 0 if it can turn off, and 100 should just be the maximum value, no?

I'm a little confused about why a percentage is being used, also -- could you clarify?

Since many commonly used digitally addressable LED channels support 256+ brightness levels, using percentage unnecessarily throws away resolution that the hardware can support. Why not at least use a u8_t, specify min and max values of 0 and 0xff in the driver API, and force the LED implementations to ignore low order bits they don't support?

That would let the users of this LED API make uniform calls, with the implementations "doing their best" in the background.

Mani-Sadhasivam replied:
Actually, the min and max percent are modelled based on the hardware supported brightness levels. For instance, in the LP3943 driver I have following code to interpret the percentage,
```
val = (value * 255) / dev_data->max_brightness;
```
The driver should take care of interpreting the brightness level based on the underlying hardware. But if this seems too confusing (more it looks like) we can switch to specifying brightness levels in `u8_t`.

mbolivar replied:
My point here is that using a percentage means only values from 0 to 100 are supported: less than 7 bits of resolution. This is a problem since a lot of hardware supports at least 8 bits of resolution. Thanks for agreeing to the change.

mbolivar commented on include/led.h:
```diff
@@ -0,0 +1,169 @@
+ *
+ * @param dev LED device
+ * @param led LED channel/pin
+ * @param period Period (in micro second) to blink an LED
```
Two issues here.

1. I still don't understand the way "period" is being used in this API.
2. I don't think "period" makes sense to use in a general API; that seems to assume PWM.

By making definitions in terms of period, this seems to assume PWM is the underlying implementation. But this is a general purpose LED API. It should be possible to implement a driver using a true DAC, which doesn't have a period: you just set brightness directly. Or to implement a driver using a GPIO, which treats any brightness other than "off" into "full on" (following the "do your best" comment in the android HAL API I linked to in my first review).

Brightness is the LED phenomenon that people perceive which is being supported in this initial version. And there's already a PWM API, so adding an LED API that assumes PWM doesn't make sense to me.

I therefore think that definitions should be done in terms of brightness levels (since the API doesn't contemplate color channels yet).

Can you try thinking through what an implementation would look like for a few different types of driver (GPIO, DAC, PWM, digitally addressable "smart" LED controlled by SPI or I2C, etc.) and see what they have in common?

Mani-Sadhasivam replied:
Most of the LED hw are capable of blinking at a specific rate, that's why I used `period` notation to specify the blinking rate per second. If this creates ambiguity with PWM, we can switch to `delay on` / `delay_off`, what do you think?

Also, I need one more clarification from you. When you say `brightness` I assume it to be the brightness level (duty cycle in PWM terms) and what about the time period? Shouldn't we add that functionality too?

mbolivar replied:
> Most of the LED hw are capable of blinking at a specific rate, that's why I used period notation to specify the blinking rate per second.

Ah, OK, so `min_period` and `max_period` were about blinking rate per second, and are related to this routine? I didn't get that; IMO it isn't clear from the docstrings ("Minimum acceptable time period (in ms) supported by the device" doesn't mention blinking).

> If this creates ambiguity with PWM, we can switch to delay on / delay_off, what do you think?

My confusion was with the period of time over which a PWM wave oscillates through a single fundamental period: the amount of time the pin is HIGH plus the time it is LOW. I was assuming this was about PWMing rapidly to support dimming, rather than blinking at perceptible rates. Make sense?

> 
> Also, I need one more clarification from you. When you say brightness I assume it to be the brightness level (duty cycle in PWM terms) and what about the time period? Shouldn't we add that functionality too? And that's what this API is intended to do.

Now I'm confused about what you're trying to do again.

Setting "brightness" with a PWM-backed LED is something you can only do if the period is small enough (say a few milliseconds) so that people see the light get brighter and darker when you adjust the duty cycle higher and lower.

But if the period is long enough (say a few seconds) so that you can perceive the light turning on and off, adjusting the duty cycle has nothing to do with brightness in people's perception.

Looks like you're not trying to support brightness after all, just blinking at rates humans can perceive?

Mani-Sadhasivam replied:
I think there is a bit of confusion here. Let me make it clear of what I'm trying to achieve with this API.

So, I want to expose functions to set both brightness and blinking of the LED as below:

1. Brightness

Most of the LED devices are capable of setting brightness at the hardware level. Here brightness refers to the level an LED is glowing. The confusion came when I used PWM notations. So, just ignore duty cycle, frequency etc... and in pure LED terms `led_set_brightness` refers to setting the brightness level at which the LED can glow constantly.

2. Blinking

This refers to turning the LED on and off at a particular rate. This doesn't have anything to do with the level of brightness an LED is glowing. So, `led_blink` will just turn on and off the LED at the specified amount of time. Here, time at which LED turns on/off is passed in `period` parameter.

In a general usecase, you can set the brightness level of an LED to say 50 (again this value is specific to the hardware) and at this brightness you can turn it on and off once every 500ms. Otherwise, you can just set the LED to maximum brightness and turn it on and off every 500ms.

If you don't set the blinking time, it will glow constantly at the specified brightness value. And in future I have plans to implement software blinking so that a general purpose timer can be used to turn the LED on/off (here LED will glow at full brightness).

mbolivar commented on include/led.h:
```diff
@@ -0,0 +1,169 @@
+ * @brief Driver specific representation of a LED device
+ */
+
+struct led_data {
```
This isn't accessible through any of the core API routines. What's going on?

Mani-Sadhasivam replied:
I intended to use this struct inside the driver specific dev_data, which means not publicly available to the user. But anyway following your comments I'm going to modify this struct to be used by the API directly.

mbolivar replied:
I looked at this again. I do think the definition of this struct ought to be moved to the LP3943 driver, since other LED implementations may not share the same constraints.

Think about a GPIO-based driver where blink is implemented in a work queue. There would then be no limit on the maximum blink time, since you can always sleep as long as necessary between toggles.

mbolivar commented on include/led.h:
```diff
@@ -0,0 +1,169 @@
+ *
+ * @see led_on() for argument descriptions.
+ */
+typedef int (*led_api_on)(struct device *dev, u32_t led);
```
Above, I proposed specifying a standard range of brightness that all drivers need to support (even if they ignore low order bits).

If that's done, then `led_on` and `led_off` can be implemented as static inlines that delegate to `led_set_brightness`:

```
static inline int led_on(struct device *dev, u32_t led)
{
	return led_set_brightness(dev, led, MAX_BRIGHTNESS);
}

static inline int led_off(struct device *dev, u32_t led)
{
	return led_set_brightness(dev, led, MIN_BRIGHTNESS);
}
```

This will save code space without restricting the API.

Mani-Sadhasivam replied:
Agree, the `led_set_brightness` API will be in following structure:
```
int led_set_brightness(struct device *dev, u32_t led, u8_t value);
```

Mani-Sadhasivam replied:
I have clarified this before during early conversation with @mbolivar. Even though we can turn off an LED by setting brightness to MIN_BRIGHTNESS, still the led would be on. The device would be consuming some current. We need to explicitly turn off the LED's and that's the whole purpose of having `led_on` and `led_off` API's.

If there are devices capable of completely turning off an LED by setting MIN_BRIGHTNESS they can ignore the above API's and implement the `led_set_brightness` alone.

mbolivar reviewed (changes requested):
Hi Mani, thanks for the update and apologies for taking so long to review it.

I think we're still not there yet in terms of the core API, which is all I've taken a look at so far.

To summarize my review:

- on and off can be implemented in terms of set_brightness, if you establish a common convention for max and min brightness

- let's leave out any API that won't get an implementation in an initial driver; it can be added later

- this still feels too much like PWM is being assumed as the backend implementation

Mani-Sadhasivam commented:
Hi @mbolivar,

Pushed another revision of the change. This includes the following changes based on your previous review comments:

1. Modified the brightness level to be represented in u8 instead of percentage.
2. Removed the unimplemented One shot blink API
3. Modified the comments to mention that the time period is used for blinking

However I kept the parameter `period` for specifying the blinking duration. It sounds simple to me rather than implementing `delay_on` and `delay_off` as in Linux LED API.

Also, I haven't implemented your recommendation of using `MAX_BRIGHTNESS` and `MIN_BRIGHTNESS` for turning an LED on/off. Reason for that is, many hardware are capable of making the LED outputs as High impedance when turned off in a proper manner. This will not be true when we use `MIN_BRIGHTNESS` to turn off an LED. IMO it should be left to the individual driver implementation without putting restriction in the LED API.

nzmichaelh commented on include/led.h:
```diff
@@ -0,0 +1,144 @@
+	/* Maximum acceptable LED blinking time period (in ms) */
+	u32_t max_period;
+	/* Minimum acceptable LED brightness value */
+	u8_t min_brightness;
```
nit: I worked on a project in the past where we needed a 10,000:1 PWM duty cycle range for a button LED at it needed to be visible in daylight and dim enough at night to not affect night vision.

Consider changing this to u16_t?

Mani-Sadhasivam replied:
Sure. u16_t sounds reasonable.

nzmichaelh commented on include/led.h:
```diff
@@ -0,0 +1,144 @@
+/**
+ * @brief Blink an LED
+ *
+ * This routine blinks an LED for the given time period continuously
```
nit: this description implies that the function blocks.  How about

"Start a LED blinking with the given period.  The LED will blink forever."

Mani-Sadhasivam replied:
Ack.

nzmichaelh commented on include/led.h:
```diff
@@ -0,0 +1,144 @@
+ *
+ * @param dev LED device
+ * @param led LED channel/pin
+ * @param period Time period (in milli seconds) an LED should blink
```
nit: s/milli seconds/milliseconds/

Mani-Sadhasivam replied:
Ack

nzmichaelh commented on include/led.h:
```diff
@@ -0,0 +1,144 @@
+/**
+ * @brief Blink an LED
+ *
+ * This routine blinks an LED for the given time period continuously
```
What duty cycle will it blink at?

Mani-Sadhasivam replied:
That depends on the value set by set_brightness. Duty cycle or brightness has to be set before calling this function. Should we need to mention it explicitly?

nzmichaelh replied:
Ah, sorry, by 'duty cycle' I mean how much of the cycle will the LED be on for.  There's a few examples at https://en.wikipedia.org/wiki/Light_characteristic including:

- Flashing (50 % on, 50 % off)
- Occulting (~20 % on, 80 % off)
- Reverse occulting (~80 % on, 20 % off)

'Flashing' SGTM.

Mani-Sadhasivam replied:
Not sure which category makes sense here. For this `led_blink` function, the device is expected to be ON for given period and OFF for the same period. For instance, if the period is 500ms then first 500ms it will be ON and next 500ms it will be OFF.

nzmichaelh commented on include/led.h:
```diff
@@ -0,0 +1,144 @@
+/**
+ * @brief Set LED brightness
+ *
+ * This routine sets the brightness of an LED to the given value
```
nit: s/an LED/a LED/

Mani-Sadhasivam replied:
Ack

nzmichaelh commented on samples/drivers/led_lp3943/src/main.c:
```diff
@@ -0,0 +1,62 @@
+
+	while (1) {
+		/* Turn on LEDs one by one */
+		for (i = 0; i <= NUM_LEDS; i++) {
```
in i <= NUM_LEDS, should this be < ?

Mani-Sadhasivam replied:
Good catch. Will fix it.

nzmichaelh commented on include/led.h:
```diff
@@ -0,0 +1,144 @@
+
+/**
+ * @typedef led_api_set_brightness()
+ * @brief Callback API for setting brightness of an LED
```
What's the relationship between 'blink' and 'set_brightness'?  As in, if I start a LED blinking, and then call set brightness, should it continue to blink?

Mani-Sadhasivam replied:
At this moment, we don't put any restrictions from API. It totally depends on how HW handles it. In future if we implement software blinking,  we may decide on cancelling blink when set_brightness is called.

Mani-Sadhasivam commented:
@mbolivar @nzmichaelh Incorporated your suggestions!

Mani-Sadhasivam commented:
@mbolivar Any update on this PR?

mbolivar commented on include/led.h:
```diff
@@ -0,0 +1,145 @@
+ * @param period Time period (in milliseconds) an LED should blink
+ * @return 0 on success, negative on error
+ */
+static inline int led_blink(struct device *dev, u32_t led, u32_t period)
```
The commit log says led_oneshot_blink() is also meant to be supported here, but I don't see it -- it's fine with me if you want to do that later, just making sure it wasn't forgotten by accident.

Mani-Sadhasivam replied:
Sorry, it is on my to-do list and not supported as of now. Will remove that from commit message.

Mani-Sadhasivam replied:
What would be the benefit of ON and OFF time than specifying the time period? Anyway we will be calculating the time period by calculating the sum of ON and OFF time in driver. IMO it will add unnecessary overhead in API.

Mani-Sadhasivam replied:
Did a bit of research and it looks like there are LED devices capable of setting the ON time and OFF time independently. Will remove `period` parameter and add `delay_on`,`delay_off`.

Thanks for pointing out :)

mbolivar commented on samples/drivers/led_lp3943/prj.conf:
```diff
@@ -0,0 +1,8 @@
+
+CONFIG_LED=y
+CONFIG_LP3943=y
+CONFIG_LP3943_I2C_MASTER_DEV_NAME="I2C_3"
```
This line is board-specific -- how about moving it to boards/96b_neonkey.conf so others can customize it to their needs in their own board files?

Mani-Sadhasivam replied:
Sure, will move it to board specific config file.

mbolivar commented on boards/arm/96b_neonkey/96b_neonkey.dts:
```diff
@@ -35,5 +35,6 @@
 
 &i2c3 {
-	clock-frequency = <I2C_BITRATE_FAST>;
+	status = "ok";
```
Is this LED driver for a chip that is built-in to the board?

If so, this seems fine, but otherwise, why not move this into a DT overlay in your sample, which  gets added to DTC_OVERLAY_FILE by the CMakeLists.txt in the sample?

Mani-Sadhasivam replied:
Yes, this chip is present on the Neonkey board.

mbolivar reviewed (commented):
@Mani-Sadhasivam, I think we are almost there. I really appreciate your honestly tremendous patience and tenacity.

I have some minor comments, then I think we should just merge this and continue discussion in tree, now that 1.12 development is open.

Mani-Sadhasivam commented:
Hi @mbolivar, 

Thanks a lot for getting back on this PR.

Mani-Sadhasivam commented:
@mbolivar So far I have addressed all of the concerns except moving `struct led_data` to device specific driver. IMO _most_ of the LED devices do need max and min values for blinking and brightness. It is obvious to provide those limitations in the API and for the devices which has no limits (like gpio-led), they can safely ignore these. Since this is a internal driver specific data, there is no visibility outside the device. Same mechanism is being used in the Linux LED API.

mbolivar commented:
> Since this is a internal driver specific data, there is no visibility outside the device.

Right, that's why they shouldn't be in include/led.h, which is for public API with visibility outside the device :).

Can you move the struct to `drivers/led/led_context.h`, the same way is done for private, driver-internal SPI APIs in `drivers/spi/spi_context.h`? That way they will still be accessible to all the LED drivers, without polluting the user-visible namespace.

Mani-Sadhasivam commented:
@mbolivar Moved `struct led_data` to `drivers/led/led_context.h`

mbolivar reviewed (approved):
I'm good with getting this API in.

Thanks a ton, @Mani-Sadhasivam!

I'm going to try to find some spare time to add color support, and versions of the existing RGB LED strip drivers which treat the entire strip as a single LED using the model you have proposed here.

Mani-Sadhasivam commented:
Thanks a lot @mbolivar for your continuous involvement from the start :) Much appreciated!

@galak @nashif Are we good to merge this?

nashif commented on drivers/led/Kconfig:
```diff
@@ -0,0 +1,38 @@
+	bool "LED drivers"
+	default n
+	help
+		Include LED drivers in the system configuration.
```
please be consistent with indentations, here it should be tab + 2 spaces.

nashif commented on drivers/led/Kconfig:
```diff
@@ -0,0 +1,38 @@
+	default 0
+	range 0 4
+	help
+		Sets the log level for LED drivers. You must have
```
ditto, tab + 2 spaces.

Mani-Sadhasivam replied:
Ack

nashif commented on drivers/led/Kconfig.lp3943:
```diff
@@ -0,0 +1,48 @@
+	depends on I2C
+	default n
+	help
+		Enable LED driver for LP3943.
```
tab + 2 spaces.

nashif commented on drivers/led/Kconfig.lp3943:
```diff
@@ -0,0 +1,48 @@
+	range 0x60 0x67
+	default 0x60
+	help
+		Specify the I2C slave address for the LP3943 LED driver.
```
tab + 2 spaces.

Mani-Sadhasivam replied:
Ack

nashif commented on drivers/led/Kconfig.lp3943:
```diff
@@ -0,0 +1,48 @@
+	string "I2C master where LP3943 is connected"
+	default "I2C_0"
+	help
+		Specify the device name of the I2C master device to which
```
ditto

Mani-Sadhasivam replied:
Ack

nashif commented on samples/drivers/led_lp3943/sample.yaml:
```diff
@@ -0,0 +1,9 @@
+tests:
+  - test:
+      platform_whitelist: 96b_neonkey
+      build_only: true
```
remove build_only prease

Mani-Sadhasivam replied:
Okay

nashif commented on samples/drivers/led_lp3943/prj.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_SYS_LOG=y
+CONFIG_BOOT_BANNER=y
```
remove this, it is added automatically.

Mani-Sadhasivam replied:
Okay

Mani-Sadhasivam commented on drivers/led/lp3943.c:
```diff
@@ -0,0 +1,319 @@
+	if (period < dev_data->min_period) {
+		period = dev_data->min_period;
+	} else if (period > dev_data->max_period) {
+		period = dev_data->max_period;
```
Okay

Mani-Sadhasivam commented on drivers/led/lp3943.c:
```diff
@@ -0,0 +1,319 @@
+
+	val &= ~(LP3943_MASK << (led << 1));
+	val |= (mode << (led << 1));
+	if (i2c_reg_write_byte(data->i2c, CONFIG_LP3943_I2C_ADDRESS,
```
Ack

Mani-Sadhasivam commented on drivers/led/lp3943.c:
```diff
@@ -0,0 +1,319 @@
+
+	val &= ~(LP3943_MASK << (led << 1));
+	val |= (mode << (led << 1));
+	if (i2c_reg_write_byte(data->i2c, CONFIG_LP3943_I2C_ADDRESS,
```
Ack

Mani-Sadhasivam commented on samples/drivers/led_lp3943/src/main.c:
```diff
@@ -0,0 +1,62 @@
+
+#define DELAY_TIME K_MSEC(1000)
+
+void main(void)
```
Will add it.

MaureenHelm replied:
Thanks

Mani-Sadhasivam commented on drivers/led/Kconfig.lp3943:
```diff
@@ -0,0 +1,48 @@
+	help
+		Device name for LP3943 LED driver.
+
+config LP3943_I2C_ADDRESS
```
Okay.

Mani-Sadhasivam commented on samples/drivers/led_lp3943/README.rst:
```diff
@@ -0,0 +1,46 @@
+Requirements
+************
+
+This sample uses the `LP3943 LED driver`_ controlled using the I2C interface.
```
This app makes use of the onboard LED's. As mentioned in the overview, each LED will be turned ON at one second interval till it reaches the end. Then it will turn OFF in reverse order.

Mani-Sadhasivam commented on samples/drivers/led_lp3943/README.rst:
```diff
@@ -0,0 +1,46 @@
+Overview
+********
+
+This sample display a continuous pattern that turns on 16 LEDs
```
Okay

Mani-Sadhasivam commented on samples/drivers/led_lp3943/README.rst:
```diff
@@ -0,0 +1,46 @@
+Building and Running
+********************
+
+Build the application for 96b_neonkey which has the LP3943 driver on board:
```
Okay

Mani-Sadhasivam commented on samples/drivers/led_lp3943/README.rst:
```diff
@@ -0,0 +1,46 @@
+   :goals: build
+   :compact:
+
+For flashing the application, refer to `Flashing`_ section of the
```
Sure

Mani-Sadhasivam commented:
@nashif @MaureenHelm @dbkinder incorporated all review comments!

Mani-Sadhasivam commented:
@MaureenHelm @nashif @dbkinder Any update on this?

MaureenHelm commented on include/led.h:
```diff
@@ -0,0 +1,133 @@
+ *
+ * @param dev LED device
+ * @param led LED channel/pin
+ * @param value Brightness value to set
```
What are the units for brightness?

MaureenHelm replied:
This would result in device-specific scales for brightness. I suggest we specify brightness as a fraction of full scale, represented in Q15 format.

Mani-Sadhasivam replied:
Okay. How do I implement it in API? Lets say that we want to pass `0.5` as the brightness value:

MaureenHelm replied:
I don't know of a general way to do this with a preprocessor macro. There are libraries to do it in runtime such as https://www.keil.com/pack/doc/CMSIS/DSP/html/group__float__to__x.html, but I wouldn't recommend using that for a constant. I suppose you could create some macros for common brightness values, such as 1.0, 0.5, 0.25.

Mani-Sadhasivam replied:
If we create macros for common values such as BRIGHT_1, BRIGHT_0_25 etc... then aren't we limiting the user in terms of granularity? What if the user wants to use 0.10 for some obvious reasons? Even the linux LED API is using the enum definitions for specifying the brightness values but I prefer to give the user full flexibility to control brightness. Then in that case a plain `uint` makes sense!

What do you think?

MaureenHelm replied:
I'm not suggesting that we use an enum in the API, because you're right, that would restrict the granularity unnecessarily. I'm suggesting that we normalize the brightness so that full scale (or max brightness) is represented by the same `u32_t` value across multiple drivers. A Q31 format would mean full scale (~1.0) is represented by the value 2^31-1=0x7FFF_FFFF. One-tenth brightness (0.1) would be represented as 2^31/10=0x0CCC_CCCC

MaureenHelm commented on drivers/led/Kconfig:
```diff
@@ -20,19 +20,27 @@ config SYS_LOG_LED_LEVEL
-		2 WARNING, write SYS_LOG_WRN in addition to previous level
-		3 INFO, write SYS_LOG_INF in addition to previous levels
-		4 DEBUG, write SYS_LOG_DBG in addition to previous levels
+	  Sets the log level for LED drivers. You must have
```
These whitespace changes need to be moved to the previous commit

MaureenHelm commented on drivers/led/lp3943.c:
```diff
@@ -0,0 +1,278 @@
+	if (value < dev_data->min_brightness) {
+		value = dev_data->min_brightness;
+	} else if (value > dev_data->max_brightness) {
+		value = dev_data->max_brightness;
```
Return an error if the requested brightness isn't supported

MaureenHelm commented on drivers/led/lp3943.c:
```diff
@@ -0,0 +1,278 @@
+	mode = LP3943_ON;
+	val &= ~(LP3943_MASK << (led << 1));
+	val |= (mode << (led << 1));
+	if (i2c_reg_write_byte(data->i2c, CONFIG_LP3943_I2C_ADDRESS,
```
Replace this call and the previous i2c call with one call to i2c_reg_update_byte

MaureenHelm commented on drivers/led/lp3943.c:
```diff
@@ -0,0 +1,278 @@
+	}
+
+	val &= ~(LP3943_MASK << (led << 1));
+	if (i2c_reg_write_byte(data->i2c, CONFIG_LP3943_I2C_ADDRESS,
```
Replace this call and the previous i2c call with one call to i2c_reg_update_byte

dbkinder commented on samples/drivers/led_lp3943/README.rst:
```diff
@@ -0,0 +1,47 @@
+Requirements
+************
+
+The :ref:`96b_neonkey`  board has an LP3943 driver on board, so we'll use
```
Ah.  I missed that this board also has the LEDs being controlled as well (so you don't need to connect an LED strip for example, as I asked).  We should mention that:

    The :ref:`96b_neonkey`  board has an LP3943 driver and 16 LEDs on board, ...

Mani-Sadhasivam replied:
okay

Mani-Sadhasivam commented:
@MaureenHelm @dbkinder Updated the PR addressing your review comments!

Mani-Sadhasivam commented:
@MaureenHelm Fixed the CI failure and all tests are passed now!
